### PR TITLE
Implement GolombRiceFilter Equality

### PR DESCRIPTION
**Pull request by nopara73**
@lontivero I used the `Data` byte array for uniqueness. That's sufficient right?

@NicolasDorier I based it on this, so it's battle tested:

https://github.com/zkSNACKs/WalletWasabi/blob/eaaef0c763200e01da0cf7de39ed74862704ffc6/WalletWasabi/Bases/ByteArraySerializableBase.cs#L55-L96

### CODE REVIEW

File: NBitcoin/BIP158/GolombRiceFilter.cs [hunk 1]
```diff
@@ -290,6 +290,38 @@ private static ulong SipHash(ulong k0, ulong k1, byte[] data)
 			hasher.Write(data);
 			return hasher.Finalize();
 		}
+
+		#region EqualityAndComparison
+
+		public override bool Equals(object obj) => obj is GolombRiceFilter tx && this == tx;
```
Comment (lontivero) [inline NBitcoin/BIP158/GolombRiceFilter.cs]: The variable name for filter shouldn't be `tx` but `filter` or any other better name.

Comment (nopara73) [reply NBitcoin/BIP158/GolombRiceFilter.cs]: fixed

**Comment by nopara73** [comment]
@lontivero Which parameters exactly?

**Comment by lontivero** [comment]
The parameters are `P` and `M`. Also, before comparing `Data` you could compare `N` (number of elements in the filter). So, if `P`, `M` and `N` are equal in both filters then it makes sense to compare the `Data` array.

**Comment by nopara73** [comment]
Done.

**Comment by NicolasDorier** [comment]
Why equality makes sense here? Why do you need it?

**Comment by nopara73** [comment]
I'm implementing something new (that honestly may never get into production) and I have equality implemented in that class, where that class has a `GolombRiceFilter` so right now I'm just ToStringing and compare equality like that.

**Comment by NicolasDorier** [comment]
I think it is counterintuitive to have an overridden equality for this class. 
This is not a value object unlike say outpoint. So I'd like to understand if there is a real use case for this or it is just a "nice to have" for some corner case reason.

**Comment by nopara73** [comment]
What do you mean? It is a value class. It maps to a filter.

**Comment by NicolasDorier** [comment]
Typically a value object map to small set of data and could potentially be a `struct` instead of a class. `Money` or `Outpoint` maps well to this concept.

Filter, not so much. What usecase do you need it, I can't see any reason you would need that.

**Comment by nopara73** [comment]
I'm writing a test and I'm asserting the serialized filter and the deserialized filter is the same.

> Filter, not so much.

Sure it could. It's not a builder or something. Maybe what confuses you is that you think about "filter" as a verb? While it's used as a noun in this context.

**Comment by NicolasDorier** [comment]
> I'm writing a test and I'm asserting the serialized filter and the deserialized filter is the same.

Is this for the only reason? Because if that was the only reason, there would be no reason to not do the same for virtually any type, like `Transaction`.

The equality in a type should be limited to what can't have any possible doubt we are comparing values and not reference. `ScriptPubKey`, `Outpoint`, `uint256` or `KeyPath` are obvious, you very often use uniqueness based on their value only and don't care about the reference. But for any other type, it is far from being obvious.

Another idea could be to have a `Filter.GetHash()` whose value is obviously comparing values.
But are you sure you want this? If it is only for a unit test, this seems pretty pointless.

If it is for a protocol, I am quite confident comparing filter equality is the wrong approach and there is a better one, so that's why I want to know why you need it.

**Comment by nopara73** [comment]
I see. I just wanted this, because comparing strings looked stupid. `.GetHash` is a good idea, I'll do that when I continue working on that branch.

**Pull request closed without merging**